File: LinkedList2.java
/**
 * Time Complexity - O(N) where N is the length of the linked list.
 * Space complexity - O(1)
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null || head.next == null)
            return null;
        
        ListNode slow = head, fast = head;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast) 
                break;
        }

        if(slow != fast)
            return null;
        
        slow = head;
        while(slow != fast){
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

File: LinkedList2.py
# Time Complexity O(n) n is the number of elements in linked list
# Space complexity O(1)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:

        slowPtr = head
        fastPtr = head

        while fastPtr and fastPtr.next:
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next.next

            if slowPtr == fastPtr:
                break
        slowPtr = head
        if fastPtr and fastPtr.next:
            while (fastPtr != slowPtr):
                slowPtr = slowPtr.next
                fastPtr = fastPtr.next
        else:
            return None

        return slowPtr
